### DISCUSSION TITLE
How to deploy RavenDB thru release pipeline in Azure Devops

### DISCUSSION THREAD

**Discussion by Kannadasa** [ravendb/ravendb#14511] [General]
I need to design a release pipeline to automate the creation of RavenDB database and future database changes using Azure Devops.  This deployment would happen in an onprem windows server or in a VM in Azure.

Any help would be highly appreciated.                                                                                                                                                      

Thanks
Kannadasan

**Comment by ayende** [comment]
RavenDB supports `xcopy` deployment.
You can just copy the binaries to the server and start it there. There are also scripts to run that will register that as a service, etc.

What specifically do you need? Also, we support docker deployments, which may be more in line with what you want.

Can you give more context?

**Reply by Kannadasa** [reply]
We need to automate the deployment/creation of a new Database using pipelines.  We are not deploying as a container so docker image would not be helpful at the moment.  We are trying to deploy RavenDB in windows server using pipelines (CI/CD)

**Reply by ayende** [reply]
What exactly are you deploying? A RavenDB instance, or a new db inside an existing instance?

**Reply by Kannadasa** [reply]
We are trying to deploy a RavenDB instance.

**Reply by ayende** [reply]
You can handle it using `xcopy` to the machine and then running the `install-service.ps` or via Terraform, see:
https://registry.terraform.io/providers/ravendb/ravendb/latest

**Reply by Kannadasa** [reply]
Ok, let me give a try using Terraform to automate the deployment

**Reply by Kannadasa** [reply]
By any chance do you know the url to download Ravendb binaries?

**Reply by ayende** [reply]
You can find them all here:

https://hibernatingrhinos.com/builds

For what you need, you'll probably use: 

https://hibernatingrhinos.com/downloads/RavenDB%20for%20Windows%20x86/53023

The number in the end is the build number.

**Reply by Kannadasa** [reply]
Do we have any other method of deploying thru pipeline instead of Terraform?

**Reply by ayende** [reply]
Can you explain what you mean `thru pipeline` ?

**Reply by Kannadasa** [reply]
We need to deploy the database using Azure Devops pipeline and any further changes to the database will be managed using release pipeline.  Basically this is automation task.  We need to automate the deployment of database and any further changes will also be automated.  Let us say we have 10 environment and we want to deploy in all the 10 environements. We setup a pipeline and mention all the 10 environments in the pipeline and define some automation mechanism (Devops) to deploy and Devops mechanism to modify the changes to database. For example if we want to modify the schema later we can modify using devops mechanism

**Reply by Kannadasa** [reply]
I am not seeing a Terraform provider for Azure in Terraform website.  I am seeing only for AWS in Terraform website and I am not sure whether Ravendb can be deployed using Terraform in on prem servers.  This url points me to AWS provider for Terraform.

https://registry.terraform.io/providers/ravendb/ravendb/1.0.1

**Reply by ayende** [reply]
The terraform provider absolutely supports Azure.
The AWS you see there is just a sample. You use Azure Terraform to setup the VM, and then use RavenDB Terrform to do the actual setup.

**Reply by Kannadasa** [reply]
Do you have any github repos for ravendb installation using Ansible and Terraform?

**Reply by ayende** [reply]
Yes: https://github.com/ravendb/terraform-provider-ravendb

**Reply by Kannadasa** [reply]
I am going to deploy Ravendb using Terraform on a local VM.

Below provider seems to be for cloud based VMs.

terraform {
  required_providers {
    ravendb = {
      source  = "ravendb/ravendb"
      version = ">=1.0.0"
    }
  }
}
    
provider "aws" {
  region = "us-east-1"
}

What provider i have to use if i have to deploy Ravendb on a local VM?

**Reply by ayende** [reply]
You need to provide the access details (ssh key and user).
Note that IIRC, we assume Ubuntu instance.

**Comment by Kannadasa** [comment]
We are going to deploy in Windows VM

**Reply by ayende** [reply]
In that case, you cannot use Terraform.
Copy the files and run the install-service.ps1

**Reply by Kannadasa** [reply]
I hope this script will create the database as well.  Please confirm.

**Reply by ayende** [reply]
No, it just deploys the server. For creating of databases, you can use REST calls or the API

**Reply by Kannadasa** [reply]
install-service.ps1 - This script is not available in the folder containing Ravendb binaries.  Do you know where can we find this script?

**Reply by ppekrol** [reply]
For linux, the file is called `install-daemon.sh`, for windows it is `setup-as-service.ps1`

**Reply by Kannadasa** [reply]
Do we have any cli commands to create a Ravendb database inside an existing instance?

**Reply by ayende** [reply]
You would need to issue a REST request for that. You can do that with CURL, see: https://ravendb.net/docs/article-page/5.4/csharp/client-api/rest-api/rest-api-intro

```
PUT /admin/databases

{
    "DatabaseName": "DB_NAME"
}
```

**Reply by Kannadasa** [reply]
If i run a curl command to create a database, I am getting the following error.  Is there anything you can help with the following error?

I have given below the commands i used and the error message

Server>curl -X PUT -H "Content-Type: application/json" \ -d "{"DatabaseName": "Kanna"}" \ http://localhost:8080/admin/databases


curl: (6) Could not resolve host: \
curl: (6) Could not resolve host: \
{"Url":"/admin/databases","Type":"System.IO.InvalidDataException","Message":"Cannot have a 'D' in this position at  (1,3) around: {DatabaseName: Kanna}","Error":"System.IO.InvalidDataException: Cannot have a 'D' in this position at  (1,3) around: {DatabaseName: Kanna}\r\n   at Sparrow.Json.Parsing.UnmanagedJsonParser.ThrowException(String message, Exception inner) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\Parsing\\UnmanagedJsonParser.cs:line 931\r\n   at Sparrow.Json.Parsing.UnmanagedJsonParser.ThrowCannotHaveCharInThisPosition(Byte b) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\Parsing\\UnmanagedJsonParser.cs:line 426\r\n   at Sparrow.Json.Parsing.UnmanagedJsonParser.ReadUnlikely(Byte b, UInt32& pos, Boolean& couldRead) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\Parsing\\UnmanagedJsonParser.cs:line 303\r\n   at Sparrow.Json.BlittableJsonDocumentBuilder.ReadMaybeModifiedPropertyName() in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\BlittableJsonDocumentBuilder.cs:line 318\r\n   at Sparrow.Json.BlittableJsonDocumentBuilder.ReadInternal[TWriteStrategy]() in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\BlittableJsonDocumentBuilder.cs:line 216\r\n   at Sparrow.Json.JsonOperationContext.ParseToMemoryAsync(Stream stream, String documentId, UsageMode mode, MemoryBuffer bytes, IBlittableDocumentModifier modifier, Nullable`1 token, Int32 maxSize) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\JsonOperationContext.cs:line 818\r\n   at Sparrow.Json.JsonOperationContext.ParseToMemoryAsync(Stream stream, String debugTag, UsageMode mode, IBlittableDocumentModifier modifier, Nullable`1 token) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\JsonOperationContext.cs:line 557\r\n   at Raven.Server.Web.System.AdminDatabasesHandler.Put() in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Raven.Server\\Web\\System\\AdminDatabasesHandler.cs:line 204\r\n   at Raven.Server.Routing.RequestRouter.HandlePath(RequestHandlerContext reqCtx) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Raven.Server\\Routing\\RequestRouter.cs:line 365\r\n   at Raven.Server.RavenServerStartup.RequestHandler(HttpContext context) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Raven.Server\\RavenServerStartup.cs:line 240"}

Thanks
Kannadasan

**Reply by krivulcik** [reply]
You are not escaping your quotes correctly.

`-d "{"DatabaseName": "Kanna"}"` becomes `-d {DatabaseName: Kanna}` (as indicated by the error message).

Either escape the value in single quotes (`-d '{"DatabaseName": "Kanna"}'`) or escape the double quotes within the outer double quotes.

**Reply by Kannadasa** [reply]
curl -X PUT -H 'Content-Type: application/json' -d '{ "DatabaseName": "DB_Kanna" }' http://localhost:8080/admin/databases

Now i am using different syntax and a different error message

curl: (6) Could not resolve host: application
curl: (3) URL using bad/illegal format or missing URL
curl: (6) Could not resolve host: DB_Kanna
curl: (3) unmatched close brace/bracket in URL position 1:
}'
 ^

**Reply by Kannadasa** [reply]
url -X PUT -H 'Content-Type: application/json' (-d '{ "DatabaseName": "DB_Kanna" }) ' http://localhost:8080/admin/databases

I tried enclosing the json objects within a bracket and this still fails

**Reply by Kannadasa** [reply]
url -X PUT -H 'Content-Type: application/json' (-d '{ "DatabaseName": "DB_Kanna" }') http://localhost:8080/admin/databases

sorry this one

**Reply by Kannadasa** [reply]
Can i use Terraform to create a  Ravendb database?

**Reply by ayende** [reply]
Yes, see: https://github.com/ravendb/terraform-provider-ravendb

```
resource "ravendb_server" "server" {
  hosts              = local.hosts
  database           = "firewire"
 ...
}

**Reply by Kannadasa** [reply]
Can i use this code to create a Ravendb database to create a database on windows VM in a on premises datacenter?

Thanks
Kannadasan

**Reply by ayende** [reply]
Probably not, our terraform setup assume that you can log in via SSH and run bash commands.

**Reply by Kannadasa** [reply]
Can i use puppet and chef to automate creation of Ravendb database on a Windows platform?

**Reply by ayende** [reply]
Yes, using curl commands or similar

**Reply by Kannadasa** [reply]
Do i need to have puppet installed on the Ravendb server ? or i can install puppet somewhere outside Ravendb server?

Thanks
Kannadasan

**Reply by ayende** [reply]
completely unrelated

**Reply by Kannadasa** [reply]
Sorry, i need to automate the changes making to Ravendb and somehow i need to track the changes with the help of some tools.  Looks like Terraform will not work to create a database if the Ravendb server is running on Windows.  Then i thought about Puppet and CHef to make changes to running database and track the changes.  If that is the case i need to install Puppet somewhere in my infrastructure to automate the creation and modification of databases using Puppet.

This is my requirement.

Thanks
Kannadasan

**Reply by ayende** [reply]
Best option is likely to just use a `powershell` script, easiest.

**Comment by Kannadasa** [comment]
Can I use some commands to create databases?

**Reply by ayende** [reply]
I don't understand your question

**Reply by Kannadasa** [reply]
public void EnsureDatabaseExists(IDocumentStore store, string database = null, bool createDatabaseIfNotExists = true)
{
    database = database ?? store.Database;

    if (string.IsNullOrWhiteSpace(database))
        throw new ArgumentException("Value cannot be null or whitespace.", nameof(database));

    try
    {
        store.Maintenance.ForDatabase(database).Send(new GetStatisticsOperation());
    }
    catch (DatabaseDoesNotExistException)
    {
        if (createDatabaseIfNotExists == false)
            throw;

        try
        {
            store.Maintenance.Server.Send(new CreateDatabaseOperation(new DatabaseRecord(database)));
        }
        catch (ConcurrencyException)
        {
            // The database was already created before calling CreateDatabaseOperation
        }

    }
}


Is this script more than enough to create a database?  I will pass this script in my pipeline as one of the tasks to create a database after "Ravendb installation" task.  I found this script in your website. (https://ravendb.net/docs/article-page/5.3/csharp/client-api/operations/server-wide/create-database#example---ensuredatabaseexists)

Task 1 :  Run the install-service.ps1 script to install the binaries.
Task 2:  Call the above script in my pipeline to create a database.

Thanks

**Reply by ayende** [reply]
Yes, that will create the database.

**Reply by Kannadasa** [reply]
We have some database already running in the server.  We dont want to disturb the existing databases.  How can we make sure that above script will not touch the existing databases and create a new databse using settings.json file

**Comment by arekpalinski** [comment]
1. Yes, that's the way to go to ensure that a database exists.
2. I believe that Oren was referring to setup-as-service.ps1 script.

On Fri, Jul 15, 2022 at 2:43 PM Kannadasa ***@***.***> wrote:

> install-service.ps1 - This script is not available in the folder
> containing Ravendb binaries. Do you know where can we find this script?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/discussions/14511#discussioncomment-3156453>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC35Z3VCMWI7PQS4OI5EVTVUFMIPANCNFSM524R3KQA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
<https://ravendb.net/> *Arkadiusz Paliński*
*Team Leader   /   Hibernating Rhinos LTD <https://hibernatingrhinos.com/>*
Sales:  ***@***.***
Support:  ***@***.***
<https://www.facebook.com/pages/RavenDB/265907650186374>
<https://twitter.com/ravendb>
<https://www.linkedin.com/company/hibernating-rhinos-ltd-/>
<https://ravendb.net/emailsignature/displayeventpage>

**Reply by Kannadasa** [reply]
Is there any way can we run this script in a non interactive mode?  Because pipeline is going to call this script and all the inputs should be available when the script is executed.

**Reply by ayende** [reply]
You need to pass it an existing `settings.json` file, yes

**Reply by Kannadasa** [reply]
Do you have any docs by which we can create a Ravendb database using cli commands by mentioning all the parameters in settings.json ?

**Reply by ayende** [reply]
There is the `rvn create-setup-package` - which outputs a zip file with everything that is required.

Honest, easiest is just to create the `settings.json` directly using text editing or `jq`

**Comment by ayende** [comment]
You were mixing " usage for the string and inside the JSON, try:

> curl -X PUT -H "Content-Type: application/json" \ -d '{"DatabaseName":
"Kanna"}' \ http://localhost:8080/admin/databases


On Wed, Jul 27, 2022 at 12:28 PM Kannadasa ***@***.***> wrote:

> If i run a curl command to create a database, I am getting the following
> error. Is there anything you can help with the following error?
>
> I have given below the commands i used and the error message
>
> Server>curl -X PUT -H "Content-Type: application/json" \ -d
> "{"DatabaseName": "Kanna"}" \ http://localhost:8080/admin/databases
>
> curl: (6) Could not resolve host:
> curl: (6) Could not resolve host:
> {"Url":"/admin/databases","Type":"System.IO.InvalidDataException","Message":"Cannot
> have a 'D' in this position at (1,3) around: {DatabaseName:
> Kanna}","Error":"System.IO.InvalidDataException: Cannot have a 'D' in this
> position at (1,3) around: {DatabaseName: Kanna}\r\n at
> Sparrow.Json.Parsing.UnmanagedJsonParser.ThrowException(String message,
> Exception inner) in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Sparrow\Json\Parsing\UnmanagedJsonParser.cs:line
> 931\r\n at
> Sparrow.Json.Parsing.UnmanagedJsonParser.ThrowCannotHaveCharInThisPosition(Byte
> b) in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Sparrow\Json\Parsing\UnmanagedJsonParser.cs:line
> 426\r\n at Sparrow.Json.Parsing.UnmanagedJsonParser.ReadUnlikely(Byte b,
> UInt32& pos, Boolean& couldRead) in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Sparrow\Json\Parsing\UnmanagedJsonParser.cs:line
> 303\r\n at
> Sparrow.Json.BlittableJsonDocumentBuilder.ReadMaybeModifiedPropertyName()
> in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Sparrow\Json\BlittableJsonDocumentBuilder.cs:line
> 318\r\n at Sparrow.Json.BlittableJsonDocumentBuilder.ReadInternal
> TWriteStrategy in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Sparrow\Json\BlittableJsonDocumentBuilder.cs:line
> 216\r\n at Sparrow.Json.JsonOperationContext.ParseToMemoryAsync(Stream
> stream, String documentId, UsageMode mode, MemoryBuffer bytes,
> IBlittableDocumentModifier modifier, Nullable1 token, Int32 maxSize) in
> C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Sparrow\\Json\\JsonOperationContext.cs:line
> 818\r\n at Sparrow.Json.JsonOperationContext.ParseToMemoryAsync(Stream
> stream, String debugTag, UsageMode mode, IBlittableDocumentModifier
> modifier, Nullable1 token) in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Sparrow\Json\JsonOperationContext.cs:line
> 557\r\n at Raven.Server.Web.System.AdminDatabasesHandler.Put() in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Raven.Server\Web\System\AdminDatabasesHandler.cs:line
> 204\r\n at
> Raven.Server.Routing.RequestRouter.HandlePath(RequestHandlerContext reqCtx)
> in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Raven.Server\Routing\RequestRouter.cs:line
> 365\r\n at Raven.Server.RavenServerStartup.RequestHandler(HttpContext
> context) in
> C:\Builds\RavenDB-Stable-5.3\53023\src\Raven.Server\RavenServerStartup.cs:line
> 240"}
>
> Thanks
> Kannadasan
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/discussions/14511#discussioncomment-3259235>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA4RM6K4JCN3ENKCTWXZ6DVWD6L7ANCNFSM524R3KQA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
<https://ravendb.net/> *Oren Eini*
*CEO   /   Hibernating Rhinos LTD <https://hibernatingrhinos.com/>*
Mobile:  972-52-548-6969
Sales:  ***@***.***
Skype:  ayenderahien
Support:  ***@***.***
<https://www.facebook.com/pages/RavenDB/265907650186374>
<https://twitter.com/ravendb>
<https://www.linkedin.com/company/hibernating-rhinos-ltd-/>
<https://ravendb.net/emailsignature/displayeventpage>

**Reply by Kannadasa** [reply]
C:\Users\kannadasan\Downloads\Ravendb\Server>curl -X PUT -H 'Content-Type: application/json' \ -d '{ "DatabaseName": "DB_Kanna" }' \ http://localhost:8080/admin/databases

curl: (6) Could not resolve host: application
curl: (6) Could not resolve host: \
curl: (3) URL using bad/illegal format or missing URL
curl: (6) Could not resolve host: DB_Kanna
curl: (3) unmatched close brace/bracket in URL position 1:
}'
 ^
I am getting error messages saying "Could not resolve host: application"

**Reply by Kannadasa** [reply]
This is the latest error message 

curl -X PUT -H "Content-Type: application/json" -d '{"DatabaseName": "DB_Kanna"}' http://localhost:8080/admin/databases

curl: (3) unmatched close brace/bracket in URL position 9:
DB_Kanna}'

**Reply by Kannadasa** [reply]
curl -X PUT -H "Content-Type: application/json" -d "{\"DatabaseName \": \"DB_Kanna\"}" http://localhost:8080/admin/databases

After fixing the escape bracket issues, I am getting the following error

{"Url":"/admin/databases","Type":"System.NullReferenceException","Message":"Object reference not set to an instance of an object.","Error":"System.NullReferenceException: Object reference not set to an instance of an object.\r\n   at Raven.Server.Web.System.AdminDatabasesHandler.Put() in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Raven.Server\\Web\\System\\AdminDatabasesHandler.cs:line 249\r\n   at Raven.Server.Routing.RequestRouter.HandlePath(RequestHandlerContext reqCtx) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Raven.Server\\Routing\\RequestRouter.cs:line 365\r\n   at Raven.Server.RavenServerStartup.RequestHandler(HttpContext context) in C:\\Builds\\RavenDB-Stable-5.3\\53023\\src\\Raven.Server\\RavenServerStartup.cs:line 240"}

**Reply by ayende** [reply]
Note here: `"DatabaseName ": `
You have an extra space in this property name.

For context, here is a command that works on the test instance:

```
 curl -X PUT -H "Content-Type: application/json" -d '{"DatabaseName": "DB_Kanna"}' http://live-test.ravendb.net/admin/databases
```

**Reply by Kannadasa** [reply]
Thanks for your help.  I am able to progress with creating database in my laptop.  But in server it fails with the following error message.

{"Type":"InvalidAuth","Message":"This server requires client certificate for authentication, but none was provided by the client. Did you forget to install the certificate?\r\n Your certificate store may be cached by the browser. Create a new private browsing tab, which will not cache any certificates. (Ctrl+Shift+N in Chrome, Ctrl+Shift+P in Firefox) Operator/ClusterAdmin access is required but not given to this certificate"}

Thanks
Kannadasan

**Reply by ayende** [reply]
https://ravendb.net/docs/article-page/5.4/csharp/client-api/rest-api/rest-api-intro#using-curl-with-https

**Reply by Kannadasa** [reply]
When i import the certificate with the private key, i get the following error message and database creation failed

url: (58) schannel: Failed to import cert file C:\......., last error is 0x80092002

Thanks
Kanadasan

**Reply by ayende** [reply]
That is just using curl, I'm not sure what command you are using, so I can't help

**Reply by Kannadasa** [reply]
Yes, I am using curl and following the syntax as provided in this url and that's when I am getting this error.
https://ravendb.net/docs/article-page/5.4/csharp/client-api/rest-api/rest-api-intro#using-curl-with-https

url: (58) schannel: Failed to import cert file C:......., last error is 0x80092002

Thanks
Kannadasan

**Reply by ayende** [reply]
What is the *full* and _exact_ command that you are running?

**Reply by Kannadasa** [reply]
curl -X PUT -H "Content-Type: application/json" -d "{\"DatabaseName\": \"DB_Name\"}" --cert "<certificate path" --key "private key path" <serverl url>

**Reply by ayende** [reply]
And the actual files you use? Are you using `crt` / `key` files or the `pfx` ?

**Reply by Kannadasa** [reply]
I am using .crt file.

I have following two paths configured in our Ravendb server.  Which one i should use in my CURL command?

D:\RavenDB\certificates\cluster.server.certificate.xxx.crt 
"D:\\RavenDB\\Node A\\Server\\cluster.server.certificate.xxx.pfx" (This path is configured in settings.json file)

I have given below the full command, I am using to import a certificate while creating a database.

curl.exe -X PUT -H "Content-Type: application/json" -d "{\"DatabaseName\": \"DB_Kanna\"}" --cert "D:\RavenDB\certificates\cluster.server.certificate.xxx.crt" --key "D:\RavenDB\certificates\cluster.server.certificate.xxx.key" https://xxx:000/admin/databases


Thanks
Kannadasan

**Reply by ayende** [reply]
`"{"DatabaseName": "DB_Kanna"}"`

You are again mixing `"` as the JSON quote style and the `"` as the command line arguments delimiter.

That is going to wreak havoc with whatever it is that `curl` is accepting.

You need to *ensure* that you use `'` for the command line arguments and `"` for the JSON.

The command should be:

```
curl.exe -X PUT -H 'Content-Type: application/json' -d '{"DatabaseName": "DB_Kanna"}' --cert 'D:\RavenDB\certificates\cluster.server.certificate.xxx.crt' --key 'D:\RavenDB\certificates\cluster.server.certificate.xxx.key' https://xxx:000/admin/databases
```

If this is still giving you an error. We need you to provide the *full* error.
That includes both the command line you executed and the full output that you got.

**Comment by Kannadasa** [comment]
I am using .crt files.

Thanks
Kannadasan

**Reply by ayende** [reply]
can you show the full command line, with all the files?

**Reply by ayende** [reply]
As well as the URL you are using.

Sending us partial information makes things a lot harder.

**Reply by Kannadasa** [reply]
Please find below the full command.

curl -X PUT -H "Content-Type: application/json" -d "{\"DatabaseName\": \"DB_Name\"}" --cert "C:\..\..\admin.client.certificate.xxx.crt" --key "C:\..\..\admin.client.certificate.xxx.key" https://xxx:443/admin/databases

Thanks
Kannadasan

**Reply by krivulcik** [reply]
Are the certificate files available for curl (and other commands) with that exact path? It's unclear what shell you use (powershell or bash/WSL), what the environment is, etc.

Also, please note that the quoting of the payload likely won't work again.

**Reply by Kannadasa** [reply]
I am running curl in the powershell in a Windows machine.  Yes certificate
files are available for curl commands.

What do you mean by :

Quoting of the payload.

 I don't understand this.

Thanks
Kannadasan

**Comment by Kannadasa** [comment]
There is no issue with the URL I think.  Error message points to
certificate files.  I will send the certificate file full path.

**Comment by Kannadasa** [comment]
I am running curl in the CMD prompt in a Windows machine.  Yes certificate
files are available for curl commands.

What do you mean by :

Quoting of the payload.

 I don't understand this.

Thanks
Kannadasan

**Comment by Kannadasa** [comment]
I have resolved the issue.  Actually curl is looking for certificate from
the certificate store with the thumbprint.  By this I was able to create a
database.

Thanks for your help.

Thanks
Kannadasan

**Comment by Kannadasa** [comment]
By using powershell script I won't be able to save the changes I am making
to a live database.

**Reply by ayende** [reply]
why not?

**Reply by Kannadasa** [reply]
can we create a client certificate using curl command in Ravendb?

**Reply by ayende** [reply]
You should do something like:

```
POST /admin/certificates

{
   "SecurityClearance": "ValidUser",
   "Name": "Cert Name",  
   "Permissions": {"DB_Name": "ReadWrite" }
}
```

**Reply by Kannadasa** [reply]
where should i mention the application name?

**Reply by ayende** [reply]
`Name`

**Comment by Kannadasa** [comment]
Can I not call this script from Terraform?

**Reply by ayende** [reply]
Sure, you can add that as a step there.
